### PR TITLE
Style Guide: enum msg, ask, tot fun, naming, meth ref, intro, #22805

### PR DESCRIPTION
**Pull request by patriknw**
Bunch of remaining sections. I think this together with previous PRs cover all (important) things from the ticket #22805, with the exception of ActorSystem initialization but that should probably be it's own topic rather than something in the style guide.

* intro
* Singleton messages, enum
* Lamdas vs. method references
* ask vs ?
* Partial vs. total Function
* Additional naming conventions

Refs #22805

### CODE REVIEW

File: akka-docs/src/main/paradox/typed/style-guide.md [hunk 1]
```diff
@@ -295,3 +300,151 @@ Scala
...
+The compiler warning if `GetValue` is not handled:
+
+```
+[warn] ... Counter.scala:45:34: match may not be exhaustive.
+[warn] It would fail on the following input: GetValue(_)
+[warn]         Behaviors.receiveMessage {
+[warn]                                  ^
+```
+
+Note that a `MatchError` will be thrown in runtime if a message is not handled, so it's important to pay
```
Comment (chbatey) [inline akka-docs/src/main/paradox/typed/style-guide.md]: at runtime?

**Review by chbatey — APPROVED** [review]
Looks good

File: akka-docs/src/main/paradox/typed/style-guide.md [hunk 2]
```diff
@@ -295,3 +300,151 @@ Scala
...
+In the `ReceiveBuilder` it can be matched in same way as other messages:
+
+Java
+:  @@snip [StyleGuideDocExamples.java](/akka-actor-typed-tests/src/test/java/jdocs/akka/typed/StyleGuideDocExamples.java) { #message-enum-match }
+
+@@@
+
+@@@ div {.group-java}
+
+## Lamdas vs. method references
```
Comment (helena) [inline akka-docs/src/main/paradox/typed/style-guide.md]: I'd spell it out for a section title
```suggestion
## Lamdas versus method references
```

File: akka-docs/src/main/paradox/typed/style-guide.md [hunk 3]
```diff
@@ -295,3 +300,151 @@ Scala
...
+
+It's recommended to keep the message matching in with the `ReceiveBuilder` as short and clean as possible
```
Comment (helena) [inline akka-docs/src/main/paradox/typed/style-guide.md]:
```suggestion
It's recommended to keep the message matching with the `ReceiveBuilder` as short and clean as possible
```

File: akka-docs/src/main/paradox/typed/style-guide.md [hunk 4]
```diff
@@ -295,3 +300,151 @@ Scala
...
+and delegate to methods. That improves readability and it's easy to navigate the methods with an IDE.
```
Comment (helena) [inline akka-docs/src/main/paradox/typed/style-guide.md]:
```suggestion
and delegate to methods. That improves readability and ease of method navigation with an IDE.
```

File: akka-docs/src/main/paradox/typed/style-guide.md [hunk 5]
```diff
@@ -295,3 +300,151 @@ Scala
...
+to lambdas can sometimes be difficult to understand.
+
+Ideally, lambdas should be written in one line of code. Two lines can be ok, but three is probably too much.
+Also, don't use braces and return statements in one-line lambda bodies.
+
+@@@
+
+@@@ div {.group-scala}
+
+## Partial vs. total Function
```
Comment (helena) [inline akka-docs/src/main/paradox/typed/style-guide.md]: Same as `vs.` thought above.

File: akka-docs/src/main/paradox/typed/style-guide.md [hunk 6]
```diff
@@ -295,3 +300,151 @@ Scala
...
+```
+
+Note that a `MatchError` will be thrown in runtime if a message is not handled, so it's important to pay
+attention to those. If a `Behavior` should not handle certain messages you can still include them
+in the pattern match and return `Behaviors.unhandled`.
+
+Scala
+:  @@snip [StyleGuideDocExamples.scala](/akka-actor-typed-tests/src/test/scala/docs/akka/typed/StyleGuideDocExamples.scala) { #pattern-match-unhandled }
+
+One thing to be aware of is that the exhaustiveness check is not enabled when there is a guard condition in the
```
Comment (helena) [inline akka-docs/src/main/paradox/typed/style-guide.md]:
```suggestion
One thing to be aware of is the exhaustiveness check is not enabled when there is a guard condition in the
```

File: akka-docs/src/main/paradox/typed/style-guide.md [hunk 7]
```diff
@@ -295,3 +300,151 @@ Scala
...
+with `Behaviors.receivePartial` or `Behaviors.receiveMessagePartial`
+
+Scala
+:  @@snip [StyleGuideDocExamples.scala](/akka-actor-typed-tests/src/test/scala/docs/akka/typed/StyleGuideDocExamples.scala) { #pattern-match-partial }
+
+@@@
+
+@@@ div {.group-scala}
+
+## ask vs ?
```
Comment (helena) [inline akka-docs/src/main/paradox/typed/style-guide.md]:
```suggestion
## `ask` versus `?`
```

File: akka-docs/src/main/paradox/typed/style-guide.md [hunk 8]
```diff
@@ -295,3 +300,151 @@ Scala
...
+
+Scala
+:  @@snip [StyleGuideDocExamples.scala](/akka-actor-typed-tests/src/test/scala/docs/akka/typed/StyleGuideDocExamples.scala) { #ask-2 }
+
+When using `?` the following doesn't compile because of type inference problem:
+
+Scala
+:  @@snip [StyleGuideDocExamples.scala](/akka-actor-typed-tests/src/test/scala/docs/akka/typed/StyleGuideDocExamples.scala) { #ask-3 }
+
+By adding parentheses it works but that is rather ugly, and therefore better to stick with `ask`.
```
Comment (helena) [inline akka-docs/src/main/paradox/typed/style-guide.md]:
```suggestion
By adding parentheses it works but is rather ugly, and therefore better to stick with `ask`.
```

File: akka-docs/src/main/paradox/typed/style-guide.md [hunk 9]
```diff
@@ -295,3 +300,151 @@ Scala
...
+
+Scala
+:  @@snip [StyleGuideDocExamples.scala](/akka-actor-typed-tests/src/test/scala/docs/akka/typed/StyleGuideDocExamples.scala) { #ask-4 }
+
+Note that `AskPattern` is only intended for request-response interaction from outside an actor. If the requester is
+inside of an actor you should prefer `ActorContext.ask` as that provides better thread-safety by not involving
```
Comment (helena) [inline akka-docs/src/main/paradox/typed/style-guide.md]:
```suggestion
inside an actor, prefer `ActorContext.ask` as it provides better thread-safety by not involving
```

**Review by helena — COMMENTED** [review]
Looking really good, but I added quite a few minor grammatical suggestions.

File: akka-docs/src/main/paradox/typed/style-guide.md (hunk 5 above)
Comment (johanandren) [reply akka-docs/src/main/paradox/typed/style-guide.md]: bit funky with entries in the TOC that disappears when you switch languages, can we make this one and the ask one L3 headers instead?

**Review by johanandren — APPROVED** [review]
LGTM with some small comments

File: akka-docs/src/main/paradox/typed/style-guide.md (hunk 2 above)
Comment (patriknw) [reply akka-docs/src/main/paradox/typed/style-guide.md]: I thought it wouldn't fit in the left ToC, but it does.
Longest is: Functional versus object-oriented style

File: akka-docs/src/main/paradox/typed/style-guide.md (hunk 7 above)
Comment (patriknw) [reply akka-docs/src/main/paradox/typed/style-guide.md]: yeah, I don't think we should have formatting in headings

File: akka-docs/src/main/paradox/typed/style-guide.md (hunk 5 above)
Comment (patriknw) [reply akka-docs/src/main/paradox/typed/style-guide.md]: what's funky about it? it works, and also I don't think normal users switch back and forth

**Comment by patriknw** [comment]
incorporated feedback, thanks, and squashed
can be merged after validation

File: akka-docs/src/main/paradox/typed/style-guide.md (hunk 2 above)
Comment (helena) [reply akka-docs/src/main/paradox/typed/style-guide.md]: Hrm, the Toc, ok.

**Pull request merged**